Spectral

pulkkins opened the pull request:
This branch contains implementations of S-PROG and STEPS in the spectral domain. These implementations have reduced computation time and memory footprint (~20-40%).

The main changes include:
- cascade.decomposition:
  - add a method for recomposing the cascade
  - add option to normalize the cascade levels
   - make computation of the cascade level statistics (mean and std) optional
  - add options to choose the input and output domains: spatial or spectral
  - in addition, implement option to use compact output in the spectral domain
   - use lowercase names for the field and mask arguments
- cascade.interface:
  - modify get_method so that two functions are returned: one for decomposing and one for recomposing the cascade
- noise.fftgenerators:
  - add option to choose the output domain for the noise: spatial or spectral
- nowcasts.sprog and nowcasts.steps:
  - add option to choose the domain for the computations: spatial or spectral
- nowcasts.utils:
  - remove computation of mean and std from stack_cascades
- timeseries.autoregression:
  - use lowercase names for variables and revise docstrings
- timeseries.correlation:
  - implement computation of correlation coefficients in the spectral domain
- utils.spectral:
  - add methods for computing mean, standard deviation and correlation coefficient in the spectral domain

pulkkins committed:
Implement nonparametric noise generation in the spectral domain

pulkkins committed:
Implement nonparametric noise generation in the spectral domain

pulkkins committed:
Add check for the domain argument

pulkkins committed:
Add check for the domain argument

pulkkins committed:
Docstring update

pulkkins committed:
Docstring update

pulkkins committed:
Add the domain argument to choose between spatial and spectral domain

pulkkins committed:
Add the domain argument to choose between spatial and spectral domain

pulkkins committed:
Add the domain option (spatial or spectral)

pulkkins committed:
Add the domain option (spatial or spectral)

pulkkins committed:
Add option to return either a stacked array or a list

pulkkins committed:
Add option to return either a stacked array or a list

pulkkins committed:
Change the name of stack_cascades and always return a list

pulkkins committed:
Change the name of stack_cascades and always return a list

pulkkins committed:
Remove unnecessary array indices

pulkkins committed:
Remove unnecessary array indices

pulkkins committed:
First version of S-PROG in the spectral domain

pulkkins committed:
First version of S-PROG in the spectral domain

pulkkins committed:
Temporary workaround for an FFT bug

pulkkins committed:
Temporary workaround for an FFT bug

pulkkins committed:
Fix incorrect normalization

pulkkins committed:
Fix incorrect normalization

pulkkins committed:
Take copies to make sure that PyFFTW does not modify the inputs

pulkkins committed:
Implement computation of correlation coefficients in the spectral domain

pulkkins committed:
Implement computation of correlation coefficients in the spectral domain

pulkkins committed:
Remove unnecessary copy

pulkkins committed:
Computation of correlation coefficients in the spectral domain

pulkkins committed:
Computation of correlation coefficients in the spectral domain

pulkkins committed:
Remove unnecessary code

pulkkins committed:
Remove unnecessary code

pulkkins committed:
Add separate options for input an and output domains

pulkkins committed:
Update the decomposition function call

pulkkins committed:
Update the decomposition function call

pulkkins committed:
Add a separate option for computing cascade level statistics

pulkkins committed:
Add a separate option for computing cascade level statistics

pulkkins committed:
Force computation of cascade statistics

pulkkins committed:
Fix dimension checking when input_domain is 'spectral'

pulkkins committed:
More general dimension checking

pulkkins committed:
More general dimension checking

pulkkins committed:
First implementation of STEPS in the spectral domain

pulkkins committed:
First implementation of STEPS in the spectral domain

pulkkins committed:
Fix incorrect spectral noise generation

pulkkins committed:
Fix incorrect spectral noise generation

pulkkins committed:
Array indexing cleanup

pulkkins committed:
Array indexing cleanup

pulkkins committed:
Fix use of incorrect fft object

pulkkins committed:
Fix use of incorrect fft object

pulkkins committed:
Merge commit

pulkkins committed:
Docstring revisions

pulkkins committed:
Fix incorrect variable names

pulkkins committed:
Use isinstance() instead of type()

pulkkins committed:
Remove commented code and unused argument

pulkkins committed:
Remove unnecessary arguments

pulkkins committed:
Move the inverse FFT outside recompose_cascade_spectral

pulkkins committed:
Rename recompose_cascade_spatial back to recompose_cascade for compatibility reasons

pulkkins committed:
Docstring revisions

pulkkins committed:
Docstring addition

pulkkins committed:
Add missing variable descriptions to docstrings

pulkkins committed:
Docstring revision

pulkkins committed:
Remove optional key-value pairs from the module docstring

pulkkins committed:
Docstring revisions

pulkkins committed:
Docstring revisions

pulkkins committed:
Fix indentation

pulkkins committed:
Implement normalization of the noise field in the spectral domain

pulkkins committed:
Add normalization of the noise field to the docstring

pulkkins committed:
Docstring revisions

pulkkins committed:
Implement computation of mean, standard deviation and correlation coefficient in the spectral domain

pulkkins committed:
Add new method names to autosummary

pulkkins committed:
Implement a method for recomposing the cascade (this will replace nowcasts.utils.recompose_cascade)

pulkkins committed:
Implement the compact_output option

pulkkins committed:
Store the input shape to the filter dictionary

pulkkins committed:
Implement computation of cascade level statistics in the spectral domain

pulkkins committed:
Add docstring and implement recomposing the cascade in the spectral domain

pulkkins committed:
Remove incorrect array stacking

pulkkins committed:
Stack weight masks and take the cascade shape from the weight mask

pulkkins committed:
Add option to normalize the cascade levels

pulkkins committed:
Set the default value of compute_stats to False

pulkkins committed:
Save normalization to the decomposition dictionary and implement normalization/renormalization correctly

pulkkins committed:
Several bugfixes for computing the decomposition in the spectral domain

pulkkins committed:
Use utils.spectral.std for computing the standard deviation of the spectral noise field

pulkkins committed:
Change the default computation domain to 'spatial'

pulkkins committed:
Make the AR(p) model more generic and use lowercase variable names

pulkkins committed:
Docstring addition

pulkkins committed:
Modify the interface so that it also returns the recomposition method

pulkkins committed:
Change error message text

pulkkins committed:
More generic implementation of temporal_autocorrelation and use lowercase variable names

pulkkins committed:
Use utils.spectral.corrcoef for computing the correlation coefficient in the spectral domain

pulkkins committed:
Use more generic indexing

pulkkins committed:
Use more generic indexing

pulkkins committed:
Add 'compact_output' to the output dictionary

pulkkins committed:
Avoid stacking the arrays when they have different shapes

pulkkins committed:
Fix typo

pulkkins committed:
Fix incorrect if clause

pulkkins committed:
More verbose error message

pulkkins committed:
Code refactoring

pulkkins committed:
Remove obsoleted methods

pulkkins committed:
Remove the option to normalize the cascade

pulkkins committed:
Remove unnecessary code

pulkkins committed:
Docstring cleanup

pulkkins committed:
Docstring correction

pulkkins committed:
Implementation of S-PROG in the spectral domain

pulkkins committed:
Fix typo

pulkkins committed:
First STEPS implementation in the spectral domain

pulkkins committed:
Code cleanup

pulkkins committed:
Allow the user to specify arbitrary time steps instead of regularly spaced ones

pulkkins committed:
Add check that the time steps are monotonously increasing

pulkkins committed:
Update the interface docstring

pulkkins committed:
Adapt the eulerian persistence and _do_nothing methods to the new interface

dnerini committed:
Fix typo

dnerini committed:
Add slack

dnerini committed:
Change PIL module import (#131)

Fix AttributeError in Job #794.4 and Job #794.2.

dnerini committed:
Change PIL module import (#131)

Fix AttributeError in Job #794.4 and Job #794.2.

pulkkins committed:
Make use of linspace compatible with NumPy 1.18

dnerini committed:
Fix handling of cases when LK returns 1 vector only

dnerini committed:
Improve docstrings

dnerini committed:
Include tapering module

dnerini committed:
Include tapering module

pulkkins committed:
Fix error message

pulkkins committed:
Docstring revisions

pulkkins committed:
Change input argument behavior

pulkkins committed:
Docstring revision

pulkkins committed:
Add missing descriptions of output keys and values

pulkkins committed:
Add more detailed description of the output cascade levels

pulkkins committed:
Use lowercase name for the cascade decomposition mask

pulkkins committed:
Docstring revision

pulkkins committed:
Docstring revisions

pulkkins committed:
Fix typo

pulkkins committed:
Raise NotImplementedError if attempting to use the SSFT noise generator in the spectral domain

pulkkins committed:
Coding style fix

pulkkins committed:
Update tests to be compatible with the new cascade decomposition interface

pulkkins committed:
Remove unnecessary input argument check

pulkkins committed:
Add check that means and stds are contained in the decomp dictionary

pulkkins committed:
Revert back to ae83020

dnerini commented:
Impressive work, @spulkkin ! I will definitely look into it and review the pull request, I hope to be able to do it in the coming days.

dnerini committed:
Merge branch 'master' into spectral

dnerini committed:
Add steps test for domain=spectral

dnerini committed:
Print domain

dnerini committed:
Print domain

dnerini committed:
Add test for sprog

dnerini committed:
Fix test

dnerini commented:
Why not having the spectral version as default domain, @spulkkin ?

dnerini committed:
Adapt sseps method

dnerini committed:
Add test for sseps method

dnerini committed:
Include main block

pulkkins commented:
@dnerini: I would not yet make "spectral" the default implementation. Looking at the S-PROG nowcasts, it produces slightly different results, although the implementations should be theoretically equivalent.

For STEPS, the verification is more difficult, as the random generators work in a different way, and it is not possible to reproduce the results using the same seed number. This could be tried with a very large ensemble (>500 members) to see if they converge to the same distribution.

Also note that the localized noise generators and sseps have not been implemented in the spectral domain. They should throw a NotImplementedError if the user chooses domain="spectral".

dnerini commented:
OK, I understand, thanks.

> Also note that the localized noise generators and sseps have not been implemented in the spectral domain. They should throw a NotImplementedError if the user chooses domain="spectral".

At the moment the sseps method doesn't accept the "domain" argument. Should I add it together with the NotImplementedError in case "spectral" is passed? Note that the interface to this method is already a bit different to the one of the steps method.

dnerini committed:
Fix interface to decomposition_fft method

dnerini commented on pysteps/nowcasts/steps.py:
```diff
@@ -137,6 +137,11 @@ def forecast(R, V, n_timesteps, n_ens_members=24, n_cascade_levels=6,
       A string defining the FFT method to use (see utils.fft.get_method).
       Defaults to 'numpy' for compatibility reasons. If pyFFTW is installed,
       the recommended method is 'pyfftw'.
+    domain : {"spatial", "spectral"}
```
Add reference to publication on spectral implementation of STEPS?

pulkkins replied:
Done.

dnerini committed:
Use utils.tapering module

dnerini reviewed (approved):
This looks good!

All tests are passing and the documentation also builds without problems:
https://pysteps.readthedocs.io/en/spectral/index.html

We are good to go!

From my side:
- updated the unit tests for the steps and sprog methods using the spectral domain
- added a new test for the sseps method
- fixed few errors caused by changes in the interface of the decomposition methods

dnerini committed:
Fix argument

pulkkins committed:
Add references to spectral implementations of S-PROG/STEPS

pulkkins committed:
Add references to the spectral methods

pulkkins commented:
> OK, I understand, thanks.
> 
> > Also note that the localized noise generators and sseps have not been implemented in the spectral domain. They should throw a NotImplementedError if the user chooses domain="spectral".
> 
> At the moment the sseps method doesn't accept the "domain" argument. Should I add it together with the NotImplementedError in case "spectral" is passed? Note that the interface to this method is already a bit different to the one of the steps method.

No need to add the "domain" argument if sseps doesn't support this functionality.

cvelascof committed:
Add importer for EUMETSAT SAF Files (#127)

* Add netCDF SAF CRRI importer

* Add test to SAF CRRI importer

* Force installing gcc-9 in travis-ci

* Run brew update-reset & brew update before installing gcc-9

- This may fix the error during gcc installation in travis-ci for osx.
- See: https://github.com/Homebrew/homebrew-core/issues/37269 and https://github.com/Homebrew/brew/issues/4755

* Update interface.py

* Sort importer methods in alphabetical order

* PEP8 changes

* Improve docstrings

Co-authored-by: Andres Perez Hortal <andresperezcba@gmail.com>
Co-authored-by: Seppo Pulkkinen <pulkkins@gmail.com>
Co-authored-by: Daniele Nerini <daniele.nerini@gmail.com>
Co-authored-by: Ruben Imhoff <31476760+RubenImhoff@users.noreply.github.com>

dnerini committed:
Fix defintion of DRMSE in docstrings

loforest commented:
Congrats for including this development!
